File: web-app/nextjs-frontend/src/types/challenge.ts
export interface ChallengeCategory {
  id: number;
  name: string;
  description: string;
}

export interface Task {
  id: number;
  title: string;
  description: string;
  order: number; // Position in the challenge
}

export interface Attachment {
  id: number;
  file_url: string;
  file_name: string;
  file_type: string; // e.g., "pdf", "image", etc.
}

export interface Rubric {
  id: number;
  criterion: string;
  description: string;
  max_score: number;
}

export interface Group {
  id: number;
  name: string;
  members: number[]; // Array of user IDs
}

export interface Prerequisite {
  id: number;
  challenge_id: number; // Reference to another challenge
  title: string;
}

export interface Feedback {
  id: number;
  content: string;
  created_by: number; // User ID
  created_at: string;
}

export interface Challenge {
  id: number;
  title: string;
  description: string;
  challenge_type: string;
  difficulty: string;
  visibility: string;
  company: number | null;
  categories: ChallengeCategory[];
  created_by: number | null;
  created_at: string;
  updated_at: string;
  start_date: string | null;
  end_date: string | null;
  duration_minutes: number | null;
  max_submissions: number;
  is_collaborative: boolean;
  max_team_size: number | null;
  skill_tags: string;
  learning_outcomes: string;
  prerequisite_description: string;
  estimated_completion_time: number | null;
  max_score: number;
  is_published: boolean;
  tasks: Task[];
  attachments: Attachment[];
  rubrics: Rubric[];
  groups: Group[];
  prerequisites: Prerequisite[];
  feedback: Feedback[];
  thumbnail: string | null;
  submission_count: number;
}

export interface CategorizedChallenges {
  [category: string]: Challenge[];
}

File: mobile-app/my_app/lib/models/challenge.dart
class Challenge {
  final int id;
  final String title;
  final String description;
  final String challengeType;
  final String difficulty;
  final String endDate;
  final List<Map<String, dynamic>> categories;
  final String? countdown;
  final int maxSubmissions;
  final bool isCollaborative;
  final int? maxTeamSize;
  final String skillTags;
  final String learningOutcomes;
  final String prerequisiteDescription;
  final int? estimatedCompletionTime;
  final double maxScore;
  final List<Map<String, dynamic>> tasks;
  final List<Map<String, dynamic>> attachments;
  final List<Map<String, dynamic>> rubrics;
  final String? thumbnail;

  Challenge({
    required this.id,
    required this.title,
    required this.description,
    required this.challengeType,
    required this.difficulty,
    required this.endDate,
    required this.categories,
    this.countdown,
    required this.maxSubmissions,
    required this.isCollaborative,
    this.maxTeamSize,
    required this.skillTags,
    required this.learningOutcomes,
    required this.prerequisiteDescription,
    this.estimatedCompletionTime,
    required this.maxScore,
    required this.tasks,
    required this.attachments,
    required this.rubrics,
    this.thumbnail,
  });

  factory Challenge.fromJson(Map<String, dynamic> json) {
    return Challenge(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      challengeType: json['challenge_type'],
      difficulty: json['difficulty'],
      endDate: json['end_date'],
      categories: List<Map<String, dynamic>>.from(json['categories']),
      countdown: json['countdown'],
      maxSubmissions: json['max_submissions'] ?? 1,
      isCollaborative: json['is_collaborative'] ?? false,
      maxTeamSize: json['max_team_size'],
      skillTags: json['skill_tags'] ?? '',
      learningOutcomes: json['learning_outcomes'] ?? '',
      prerequisiteDescription: json['prerequisite_description'] ?? '',
      estimatedCompletionTime: json['estimated_completion_time'],
      maxScore: (json['max_score'] as num?)?.toDouble() ?? 100.0,
      tasks: List<Map<String, dynamic>>.from(json['tasks'] ?? []),
      attachments: List<Map<String, dynamic>>.from(json['attachments'] ?? []),
      rubrics: List<Map<String, dynamic>>.from(json['rubrics'] ?? []),
      thumbnail: json['thumbnail'],
    );
  }
}